content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,209 @@ Like a segment tree on values rather than indices.
+prefixB[i] = prefixB[i - 1] + c[i]
+$$
+
+### Updating 𝑟 as We Recurse
+
+To know the value of 𝑟 if we recurse left, we use prefixB[r]
+If we recurse right, we use 𝑟 - prefixB[r]
+
+## Solving the second type of query (Given a range l, r find the k smallest element)
+
+We will determine whether the answer for a given node is in the left or the
+right segment.
+We can calculate how many times the elements within the segments' ranges appear
+in our range (l, r) using our first type of query.
+Note that this also works for non-leaf nodes using the following formula:
+
+$$
+occurrences(l, r) = r - l
+$$
+<Info title="Simular">
+This is simular to counting how many times a value appears up to index 𝑅 in our previous query. We did this by using the new 𝑅 value at the leaf node. But now, we consider the difference between the updated 𝑅 and 𝐿
+</Info>
+
+Therefore, the occurrences of the left node is
+
+$$
+left_occurrences = prefixB[r] - prefixB[l]
+$$
+
+<Info title="Left occurrences">
+  Note that left_occurrences is the amount of indices between l and r that their value is less than mid
+
+</Info>
+
+- If left_occurrences left_occurrences is greater or equal to k, it means the k-th smallest element is in
+  the left subtree. Therefore, we update our range and recurse into the left
+  child
+- If left_occurrences is less than k, it means the
+  k-th smallest element is in the right subtree. We adjust k by subtracting
+  left_occurrences from k, update our range, and recurse into the right child
+
+<Info title="Notice">
+  Notice we still update l, r accordingly when we go left or right
+</Info>
+
+the answer then will be the value of the node we end up on (leaf)
+
+## implemention
+
+<CPPSection>
+
+```cpp
+struct Segment {
+	Segment *left = nullptr, *right = nullptr;
+	int l, r, mid;
+	bool children = false;
+	vector<pair<int, int>> indices;  // Index, Value
+	vector<int> prefixB;
+
+	Segment(int l, int r, vector<pair<int, int>> indices)
+	    : l(l), r(r), mid((r + l) / 2), indices(std::move(indices)) {
+		calculateB();
+	}
+
+	void update() {
+		if (children) return;
+		children = true;
+		if (r - l > 1) {
+			// Split the indices for left and right child
+			vector<pair<int, int>> leftIndices, rightIndices;
+			for (auto [ind, val] : indices)
+				if (val < mid) leftIndices.emplace_back(ind, val);
+				else rightIndices.emplace_back(ind, val);
+
+			left = new Segment(l, mid, leftIndices);
+			right = new Segment(mid, r, rightIndices);
+		}
+	}
+
+	// Calculates the prefix B
+	void calculateB() {
+		int i = 1;
+		int j = 0;
+		prefixB.resize(indices.size() + 1);
+		for (auto [ind, val] : indices) {
+			if (val < mid) j++;
+			prefixB[i++] = j;
+		}
+	}
+
+	int findKSmallest(int a, int b, int k) {
+		update();
+		if (r - l <= 1) return l;
+
+		int lb = prefixB[a];
+		int lr = prefixB[b];
+		int inLeft = lr - lb;
+
+		if (k <= inLeft) return left->findKSmallest(lb, lr, k);
+		else return right->findKSmallest(a - lb, b - lr, k - inLeft);
+	}
+};
```

Comment:
no main method?

Reply:
Could be added, the main idea is the implementation of the Wavelet Tree.
(Added)

Reply:
is there a problem we can submit to to test the basic functionality of this

Reply:
Yes. I have added the main for the focus problem.